src/backend/drm/device/mod.rs

```diff
@@ -0,0 +1,586 @@
+                dev: self.dev_path(),
+                source,
+            })?;
+            let filter = info.possible_crtcs();
+            if self.resources.filter_crtcs(filter).contains(crtc) {
+                match self.plane_type(*plane)? {
+                    PlaneType::Primary => {
+                        primary = Some(*plane);
+                    }
+                    PlaneType::Cursor => {
+                        cursor = Some(*plane);
+                    }
+                    PlaneType::Overlay => {
+                        overlay.push(*plane);
+                    }
+                };
+            }
+        }
+
+        Ok(Planes {
+            primary: primary.expect("Crtc has no primary plane"),
+            cursor,
+            overlay: if self.has_universal_planes {
+                Some(overlay)
+            } else {
+                None
+            },
+        })
+    }
+
+    fn plane_type(&self, plane: plane::Handle) -> Result<PlaneType, Error> {
+        let props = self.get_properties(plane).map_err(|source| Error::Access {
+            errmsg: "Failed to get properties of plane",
+            dev: self.dev_path(),
+            source,
+        })?;
+        let (ids, vals) = props.as_props_and_values();
+        for (&id, &val) in ids.iter().zip(vals.iter()) {
+            let info = self.get_property(id).map_err(|source| Error::Access {
+                errmsg: "Failed to get property info",
+                dev: self.dev_path(),
+                source,
+            })?;
+            if info.name().to_str().map(|x| x == "type").unwrap_or(false) {
+                return Ok(match val {
+                    x if x == (PlaneType::Primary as u64) => PlaneType::Primary,
+                    x if x == (PlaneType::Cursor as u64) => PlaneType::Cursor,
+                    _ => PlaneType::Overlay,
+                });
+            }
+        }
+        unreachable!()
+    }
+
+    /// Creates a new rendering surface.
+    ///
+    /// # Arguments
+    ///
+    /// Initialization of surfaces happens through the types provided by
+    /// [`drm-rs`](drm).
+    ///
+    /// - [`crtc`](drm::control::crtc)s represent scanout engines of the device pointing to one framebuffer. \
+    ///     Their responsibility is to read the data of the framebuffer and export it into an "Encoder". \
+    ///     The number of crtc's represent the number of independant output devices the hardware may handle.
+    /// - [`plane`](drm::control::plane)s represent a single plane on a crtc, which is composite together with
+    ///     other planes on the same crtc to present the final image.
+    /// - [`mode`](drm::control::Mode) describes the resolution and rate of images produced by the crtc and \
+    ///     has to be compatible with the provided `connectors`.
+    /// - [`connectors`] - List of connectors driven by the crtc. At least one(!) connector needs to be \
+    ///     attached to a crtc in smithay.
+    pub fn create_surface(
```

Comment:
So, from this doc I gather that if one is to use multiple planes, then a `DrmSurface` needs to be created for each.

Does the user have to provide the same set of `connector`s each time? What happens if multiple surfaces are created one the same crtc for different planes but a different set of connectors is provided each time?

And btw, what happens if I try to create the same surface (so for the same pair of crtc+plane) multiple times?

Reply:
Oh yeah, that is a good point. This is not only lacking on a documentation level, but probably also on an API level.

> So, from this doc I gather that if one is to use multiple planes, then a `DrmSurface` needs to be created for each.

yes.

> Does the user have to provide the same set of `connector`s each time? What happens if multiple surfaces are created one the same crtc for different planes but a different set of connectors is provided each time?

Different planes on the same crtc with the same connector will simply not work, which also means adding and removing connectors will be problematic, when you have multiple planes, because there is no way to commit them together and the atomic check would likely always fail.

This needs an API redesign. We probably need just one surface per crtc and another `plane` type that can be used e.g. for rendering to the different planes, but has nothing todo with managing surface-global stuff like connectors.
(Which makes the `DrmRenderSurface` an even worse abstraction, because that could not really handle multiple planes at all.)

> And btw, what happens if I try to create the same surface (so for the same pair of crtc+plane) multiple times?

The Creation for the second one should also succeed and modify the drm state without updating the old one of course. So they would both probably be able to commit and alter the state to some degree, but stumble upon each other.

And I think I do not really want to fix this, because that is the user doing stupid stuff. I mean we could track used crtc/plane combinations inside the device, but I will probably just document that this may cause undefined behavior. What do you think?

Reply:
Should be fixed by a4f66da.